plutus-contract/test/Spec/TxConstraints/MustSpendScriptOutput.hs

```diff
@@ -85,338 +130,593 @@ mustSpendScriptOutputsContract a b c = mustSpendScriptOutputsContract' a b c Tru
+    checkPredicate
     "Successful use of mustSpendScriptOutput for all script's UtxOs"
     (valueAtAddress someAddress (== Ada.lovelaceValueOf 0)
     .&&. assertValidatedTransactionCount 2)
-    (void $ trace $ mustSpendScriptOutputsContract PlutusV1 5 5)
+    (void $ trace $ mustSpendScriptOutputsContract l 5 5)
 
 -- | Uses onchain and offchain constraint mustSpendScriptOutput to spend some of the UtxOs locked
 -- by the script
-validUseOfMustSpendScriptOutputUsingSomeScriptOutputs :: TestTree
-validUseOfMustSpendScriptOutputUsingSomeScriptOutputs =
-    checkPredicateOptions
-    defaultCheckOptions
+validUseOfMustSpendScriptOutputUsingSomeScriptOutputs :: PSU.Language -> TestTree
+validUseOfMustSpendScriptOutputUsingSomeScriptOutputs l =
+    checkPredicate
     "Successful use of mustSpendScriptOutput for some of the script's UtxOs"
     (valueAtAddress someAddress (== V.scale 2 utxoValue)
     .&&. assertValidatedTransactionCount 2)
-    (void $ trace $ mustSpendScriptOutputsContract PlutusV1 5 3)
+    (void $ trace $ mustSpendScriptOutputsContract l 5 3)
+
+-- | Uses onchain and offchain constraint mustSpendScriptOutputWithReference to spend some of the UtxOs locked
+-- by the script
+validUseOfReferenceScript :: PSU.Language -> TestTree
+validUseOfReferenceScript l = let
+    contract = mustSpendScriptOutputWithReferenceContract l 1 True
+    versionedMintingPolicy = getVersionedScript MustSpendScriptOutputWithReferencePolicy l
+    check = case l of
+        PlutusV1 ->
+            checkPredicateOptions
+            (changeInitialWalletValue w1 (const $ Ada.adaValueOf 1000) defaultCheckOptions)
+            "Phase 1 validation error when we used Reference script in a PlutusV1 script"
+            ( assertFailedTransaction ( const $ has
+                $ L._CardanoLedgerValidationError . filtered (Text.isPrefixOf "ReferenceInputsNotSupported")
+            ) .&&. assertValidatedTransactionCountOfTotal 1 2
+            )
+        PlutusV2 ->
+            checkPredicateOptions
+            (changeInitialWalletValue w1 (const $ Ada.adaValueOf 1000) defaultCheckOptions)
+            "Successful use of mustSpendScriptOutputWithReference to unlock funds in a PlutusV2 script"
+            (walletFundsChange w1 (tokenValue versionedMintingPolicy)
+            .&&. valueAtAddress (scriptAddress MustReferenceOutputValidator l ) (== Ada.adaValueOf 0)
+            .&&. assertValidatedTransactionCount 2
+            )
+    in check $ traceN 3 contract
+
+-- | Uses onchain and offchain constraint mustSpendScriptOutputWithReference several time on the same script to spend
+-- some of the UtxOs locked by the script
+validMultipleUseOfTheSameReferenceScript :: PSU.Language -> TestTree
+validMultipleUseOfTheSameReferenceScript l = let
+    contract = mustSpendScriptOutputWithReferenceContract l 2 True
+    versionedMintingPolicy = getVersionedScript MustSpendScriptOutputWithReferencePolicy l
+    check = case l of
+        PlutusV1 ->
```

Comment:
Is it worth performing this check for every test?

In other suites we've done this by having 1 test to check the version incompatibility, such as https://github.com/input-output-hk/plutus-apps/blob/5ca420e93c2616aa2efb14db7d682e526427c172/plutus-contract/test/Spec/TxConstraints/MustPayToOtherScript.hs#L77-L79

Reply:
You're right, especially as we always failed with the same error.